File: android/app/src/test/java/com/hawksnest/core/logic/LockCodesTest.kt
package com.hawksnest.core.logic

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

/** Ported from `src/lib/__tests__/lockCodes.test.ts`. */
class LockCodesTest {

    @Test
    fun `validates 4 to 8 digit codes`() {
        assertTrue(isValidUserCode("1234"))
        assertTrue(isValidUserCode("12345678"))
        assertFalse(isValidUserCode("123"))
        assertFalse(isValidUserCode("123456789"))
        assertFalse(isValidUserCode("12a4"))
        assertFalse(isValidUserCode(""))
    }

    @Test
    fun `guest automation ids are stable and recognizable`() {
        val id = guestAutomationId("lock.front_door_lock", 3)
        assertEquals("hawksnest_guest_lock_front_door_lock_slot3", id)
        assertTrue(isGuestAutomation(id))
        assertFalse(isGuestAutomation("1700000000000"))
        assertEquals(3, guestSlotFromId(id))
        assertNull(guestSlotFromId("1700000000000"))
    }

    @Test
    fun `lock entity detection`() {
        assertTrue(isLockEntity("lock.front_door_lock"))
        assertFalse(isLockEntity("light.basement"))
    }

    @Test
    fun `guest expiry automation is null for invalid datetime`() {
        assertNull(buildGuestExpiryAutomation("lock.front_door_lock", 3, "Sitter", "nope"))
    }

    @Test
    fun `guest expiry automation has the expected structure`() {
        val cfg = buildGuestExpiryAutomation("lock.front_door_lock", 4, "Sitter", "2026-07-01T14:30")!!
        assertEquals("hawksnest_guest_lock_front_door_lock_slot4", cfg["id"]!!.jsonPrimitive.content)

        val trigger = (cfg["trigger"] as JsonArray)[0].jsonObject
        assertEquals("time", trigger["platform"]!!.jsonPrimitive.content)
        assertEquals("14:30:00", trigger["at"]!!.jsonPrimitive.content)

        val condition = (cfg["condition"] as JsonArray)[0].jsonObject
        assertEquals(
            "{{ as_timestamp(now()) >= as_timestamp('2026-07-01 14:30:00') }}",
            condition["value_template"]!!.jsonPrimitive.content,
        )

        val actions = cfg["action"] as JsonArray
        val clear = actions[0].jsonObject
        assertEquals("zwave_js.clear_lock_usercode", clear["service"]!!.jsonPrimitive.content)
        assertEquals(4, (clear["data"] as JsonObject)["code_slot"]!!.jsonPrimitive.content.toInt())
        assertEquals(
            "lock.front_door_lock",
            (clear["target"] as JsonObject)["entity_id"]!!.jsonPrimitive.content,
        )
        assertEquals("automation.turn_off", actions[1].jsonObject["service"]!!.jsonPrimitive.content)
    }
}

File: src/lib/__tests__/lockCodes.test.ts
import { describe, it, expect } from "vitest";
import {
  isValidUserCode,
  setUserCodeCall,
  clearUserCodeCall,
  guestAutomationId,
  isGuestAutomation,
  buildGuestExpiryAutomation,
  isLockEntity,
} from "../lockCodes";

describe("isValidUserCode", () => {
  it("accepts 4–8 digit codes", () => {
    expect(isValidUserCode("1234")).toBe(true);
    expect(isValidUserCode("12345678")).toBe(true);
  });
  it("rejects too short / too long / non-numeric", () => {
    expect(isValidUserCode("123")).toBe(false);
    expect(isValidUserCode("123456789")).toBe(false);
    expect(isValidUserCode("12a4")).toBe(false);
    expect(isValidUserCode("")).toBe(false);
  });
});

describe("service calls", () => {
  it("builds set/clear usercode calls", () => {
    expect(setUserCodeCall(3, "4242")).toEqual({
      domain: "zwave_js",
      service: "set_lock_usercode",
      data: { code_slot: 3, usercode: "4242" },
    });
    expect(clearUserCodeCall(3)).toEqual({
      domain: "zwave_js",
      service: "clear_lock_usercode",
      data: { code_slot: 3 },
    });
  });
});

describe("guest automation ids", () => {
  it("is stable and recognizable per lock+slot", () => {
    const id = guestAutomationId("lock.front_door_lock", 3);
    expect(id).toBe("hawksnest_guest_lock_front_door_lock_slot3");
    expect(isGuestAutomation(id)).toBe(true);
    expect(isGuestAutomation("1700000000000")).toBe(false);
  });
});

describe("buildGuestExpiryAutomation", () => {
  it("returns null for an invalid datetime-local value", () => {
    expect(
      buildGuestExpiryAutomation({
        lockEntityId: "lock.front_door_lock",
        slot: 3,
        guestName: "Sitter",
        expiryLocal: "not-a-date",
      }),
    ).toBeNull();
  });

  it("builds a daily time trigger gated by a datetime template, clearing then self-disabling", () => {
    const cfg = buildGuestExpiryAutomation({
      lockEntityId: "lock.front_door_lock",
      slot: 4,
      guestName: "Sitter",
      expiryLocal: "2026-07-01T14:30",
    })!;
    expect(cfg.id).toBe("hawksnest_guest_lock_front_door_lock_slot4");
    expect(cfg.trigger).toEqual([{ platform: "time", at: "14:30:00" }]);
    expect(cfg.condition).toEqual([
      {
        condition: "template",
        value_template:
          "{{ as_timestamp(now()) >= as_timestamp('2026-07-01 14:30:00') }}",
      },
    ]);
    const actions = cfg.action as Array<Record<string, unknown>>;
    expect(actions[0]).toEqual({
      service: "zwave_js.clear_lock_usercode",
      target: { entity_id: "lock.front_door_lock" },
      data: { code_slot: 4 },
    });
    expect(actions[1]).toMatchObject({ service: "automation.turn_off" });
  });
});

describe("isLockEntity", () => {
  it("recognizes lock entities only", () => {
    expect(isLockEntity("lock.front_door_lock")).toBe(true);
    expect(isLockEntity("light.basement")).toBe(false);
  });
});
